src/OneTimeTokenProvider.js

```diff
@@ -0,0 +1,43 @@
+import Connection from './Connection'
+
+export default class OneTimeTokenProvider {
+  constructor(publicAPIKey, options) {
+    this.publicAPIKey = publicAPIKey
+    this.options = options
+    this.token = null
```

Comment:
I wonder if `this.token` and `this.timeout` would fit well into a `OneTimeToken` class. `OneTimeToken` could handle its validity and timeout by itself. And `OneTimeTokenProvider` would only take care of providing a valid `OneTimeToken`.

Reply:
Thank you for the suggestion 😃
I like such as the idea of using a value object.

But it's hard to imagine how we should deal with the cache and its expiration for me a little bit.
For example, what should own the cache? If it is `OneTimeTokenProvider`, and `OneTimeToken` should expire itself, how do the token clear the cache? In this case, I think `OneTimeToken` is owned by `OneTimeTokenProvider`, so `OneTimeToken` shouldn't have a reference to the provider.

Could you share more details?

Reply:
I think in such a case `OneTimeToken` could have a flag `hasExpired()` if it has expired or not (-> the token handles expiration). And it could have a method to cause immediate expiration `expireNow()` (which would stop the timers and set the expiration flag).

And how about: `OneTimeTokenProvider` owns a `OneTimeToken`. Therefore `OneTimeTokenProvider` is responsible for handling/invalidating the cache/storing a token.

We could check if `someToken.hasExpired()` is true, to invalidate the cache and create a new token object. And when someone uses `oneTimeTokenProvider.get()` we could call `someToken.expireNow()` to expire the token immediately because it was used.

Hmm.. though now I'm not sure. It's not easy.. I'm happy to hear your thoughts.

Reply:
Did you mean a current-time-based logic instead of `setTimeout` by your suggestion?
In the current architecture, there is no special reason to pack a token value and expiration information, because `expires_in` is used only from `OneTimeTokenProvider`.
Considering to remove `setTimeout`, it would be beneficial because the timing to execute the callback is unstable, and complexity of asynchronous programming.
In conclusion, I think both would work, but to simplify, we can refactor by above reason.
How about this aspect?

Reply:
Considering the use of the token, I think you are right.

Ah, I didn't intend to suggest to remove `setTimeout`. But now that you said it, I think it's a good idea. I agree with your reasons.

Reply:
OK, thank you for the discussion! I'll apply the changes 👍